File: engines/web/src/catch-network-partition.test.ts
// Copyright (c) 2026 Dennis Brandl
// Licensed under the Apache License, Version 2.0. See LICENSE for details.
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import { partitionCatchNetworks } from './catch-network-partition.js';

const steps = [
  { oid: 's1', step_type: 'START' },
  { oid: 's2', step_type: 'ACTION PROXY' },
  { oid: 's3', step_type: 'END' },
  { oid: 'c1', step_type: 'CATCH', catch_id: 'C1' },
  { oid: 'c2', step_type: 'USER_INTERACTION' },
  { oid: 'c3', step_type: 'RETURN' },
];
const connections = [
  { from_step_id: 's1', to_step_id: 's2' },
  { from_step_id: 's2', to_step_id: 's3' },
  { from_step_id: 'c1', to_step_id: 'c2' },
  { from_step_id: 'c2', to_step_id: 'c3' },
];

describe('partitionCatchNetworks', () => {
  it('separates main-flow from catch-network steps', () => {
    const p = partitionCatchNetworks(steps, connections);
    assert.deepEqual([...p.mainFlowStepOids].sort(), ['s1', 's2', 's3']);
    assert.deepEqual([...p.catchNetworkStepOids].sort(), ['c1', 'c2', 'c3']);
  });

  it('indexes networks by catch_id', () => {
    const p = partitionCatchNetworks(steps, connections);
    assert.deepEqual([...(p.networksByCatchId.get('C1') ?? [])].sort(), ['c1', 'c2', 'c3']);
  });

  it('excludes a CATCH whose downstream has no RETURN', () => {
    const p = partitionCatchNetworks(
      [{ oid: 's1', step_type: 'START' }, { oid: 's2', step_type: 'END' },
       { oid: 'c1', step_type: 'CATCH', catch_id: 'C1' }, { oid: 'c2', step_type: 'USER_INTERACTION' }],
      [{ from_step_id: 's1', to_step_id: 's2' }, { from_step_id: 'c1', to_step_id: 'c2' }],
    );
    assert.equal(p.catchNetworkStepOids.size, 0);
    assert.equal(p.mainFlowStepOids.size, 4);
  });
});

File: engines/kmp-engine/src/jvmTest/kotlin/com/trajectoryruntime/engine/CatchNetworkPartitionTest.kt
// Copyright (c) 2026 Dennis Brandl
// Licensed under the Apache License, Version 2.0. See LICENSE for details.
package com.trajectoryruntime.engine

import kotlin.test.Test
import kotlin.test.assertEquals

class CatchNetworkPartitionTest {
  private val steps = listOf(
    Triple("s1", "START", null), Triple("s2", "ACTION PROXY", null), Triple("s3", "END", null),
    Triple("c1", "CATCH", "C1"), Triple("c2", "USER_INTERACTION", null), Triple("c3", "RETURN", null),
  )
  private val conns = listOf("s1" to "s2", "s2" to "s3", "c1" to "c2", "c2" to "c3")

  @Test fun `separates main-flow from catch-network`() {
    val p = partitionCatchNetworks(steps, conns)
    assertEquals(setOf("s1", "s2", "s3"), p.mainFlowStepOids)
    assertEquals(setOf("c1", "c2", "c3"), p.catchNetworkStepOids)
    assertEquals(setOf("c1", "c2", "c3"), p.networksByCatchId["C1"])
  }

  @Test fun `excludes a CATCH with no reachable RETURN`() {
    val p = partitionCatchNetworks(
      listOf(Triple("s1","START",null), Triple("s2","END",null), Triple("c1","CATCH","C1"), Triple("c2","USER_INTERACTION",null)),
      listOf("s1" to "s2", "c1" to "c2"),
    )
    assertEquals(0, p.catchNetworkStepOids.size)
    assertEquals(4, p.mainFlowStepOids.size)
  }
}
